### PR TITLE
Handle invalid calls, add checks (guards)

### PR DESCRIPTION
**Pull request by andreibancioiu**

### CODE REVIEW

File: arwen/context/arwen.go [hunk 1]
```diff
@@ -303,6 +303,22 @@ func (host *vmContext) createVMOutputInCaseOfError(errCode vmcommon.ReturnCode)
 	return vmOutput
 }
 
+func (host *vmContext) getFunctionToCall() (func(...interface{}) (wasmer.Value, error), error) {
+	exports := host.instance.Exports
+
```
Comment (sasurobert) [inline arwen/context/arwen.go]: delete empty line

File: arwen/context/arwen.go [hunk 2]
```diff
@@ -303,6 +303,22 @@ func (host *vmContext) createVMOutputInCaseOfError(errCode vmcommon.ReturnCode)
...
+	function, ok := exports[host.callFunction]
+
```
Comment (sasurobert) [inline arwen/context/arwen.go]: delete empty line

File: arwen/context/arwen.go [hunk 3]
```diff
@@ -974,7 +990,9 @@ func (host *vmContext) createETHCallInput() []byte {
 	}
 
 	for _, arg := range host.vmInput.Arguments {
-		newInput = append(newInput, arg...)
+		paddedArg := make([]byte, arwen.HashLen)
```
Comment (sasurobert) [inline arwen/context/arwen.go]: maybe add a new constant to EthereumArgumentLen

File: arwen/crypto/cryptoei.go [hunk 4]
```diff
@@ -76,3 +82,16 @@ func keccak256(context unsafe.Pointer, dataOffset int32, length int32, resultOff
...
+		instCtx := wasmer.IntoInstanceContext(context)
+		hostContext := arwen.GetCryptoContext(instCtx.Data())
+		hostContext.SignalUserError()
+		hostContext.UseGas(hostContext.GasLeft())
+
+		return true
+	}
+
+	return false
+}
```
Comment (sasurobert) [inline arwen/crypto/cryptoei.go]: add empty line after. need an empty line at the end of file.

File: arwen/elrondapi/elrondei.go [hunk 5]
```diff
@@ -466,11 +501,21 @@ func writeLog(context unsafe.Pointer, pointer int32, length int32, topicPtr int3
...
+
+	topics, err := arwen.GuardedMakeByteSlice2D(numTopics)
+	if withFault(err, context) {
+		return
+	}
 
-	topics := make([][]byte, numTopics)
 	for i := int32(0); i < numTopics; i++ {
-		topics[i] = arwen.LoadBytes(instCtx.Memory(), topicPtr+i*arwen.HashLen, arwen.HashLen)
+		topics[i], err = arwen.LoadBytes(instCtx.Memory(), topicPtr+i*arwen.HashLen, arwen.HashLen)
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go]: if we change LoadBytes and StoreBytes to get the context, than fault handling can be done inside.

File: arwen/elrondapi/elrondei.go [hunk 6]
```diff
@@ -973,10 +1089,27 @@ func getReturnData(context unsafe.Pointer, resultId int32, dataOffset int32) int
...
+		instCtx := wasmer.IntoInstanceContext(context)
+		hostContext := arwen.GetErdContext(instCtx.Data())
+		hostContext.SignalUserError()
+		hostContext.UseGas(hostContext.GasLeft())
+
+		return true
+	}
+
+	return false
+}
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go]: add an empty line at the end of file.

File: arwen/ethapi/ethereumei.go [hunk 7]
```diff
@@ -760,7 +895,9 @@ func ethcallStatic(context unsafe.Pointer, gasLimit int64, addressOffset int32,
 		RecipientAddr: address,
 		Function:      "main",
 	}
-	err := ethContext.ExecuteOnSameContext(contractCallInput)
+
+	err = ethContext.ExecuteOnSameContext(contractCallInput)
+
```
Comment (sasurobert) [inline arwen/ethapi/ethereumei.go]: delete empty line

File: arwen/ethapi/ethereumei.go [hunk 8]
```diff
@@ -760,7 +895,9 @@ func ethcallStatic(context unsafe.Pointer, gasLimit int64, addressOffset int32,
 		RecipientAddr: address,
 		Function:      "main",
 	}
-	err := ethContext.ExecuteOnSameContext(contractCallInput)
+
```
Comment (sasurobert) [inline arwen/ethapi/ethereumei.go]: delete empty line

File: arwen/ethapi/ethereumei.go [hunk 9]
```diff
@@ -822,4 +970,17 @@ func converToEthU128(data []byte) []byte {
...
+		instCtx := wasmer.IntoInstanceContext(context)
+		hostContext := arwen.GetEthContext(instCtx.Data())
+		hostContext.SignalUserError()
+		hostContext.UseGas(hostContext.GasLeft())
+
+		return true
+	}
+
+	return false
 }
```
Comment (sasurobert) [inline arwen/ethapi/ethereumei.go]: add empty line to the end of file.

File: arwen/context/arwen.go (hunk 3 above)
Comment (andreibancioiu) [reply arwen/context/arwen.go]: Called it `ArgumentLenEth`, to be consistent with the other constants in `helpers.go`.

File: arwen/elrondapi/elrondei.go (hunk 5 above)
Comment (andreibancioiu) [reply arwen/elrondapi/elrondei.go]: It is somehow possible, but we need a similar check in the API function anyway, in order to end execution on fault, with `return`:

```
if ... {
   return ... (end API function execution)
}
```

**Pull request merged**